observer/probers/tls/tls.go

```diff
@@ -0,0 +1,193 @@
+	reason     *prometheus.GaugeVec
+}
+
+// Name returns a string that uniquely identifies the monitor.
+func (p TLSProbe) Name() string {
+	return fmt.Sprintf("%s-expecting-%s-%s-%s", p.url, p.response, p.rootOrg, p.rootCN)
+}
+
+// Kind returns a name that uniquely identifies the `Kind` of `Prober`.
+func (p TLSProbe) Kind() string {
+	return "TLS"
+}
+
+// Get OCSP status (good, revoked or unknown) of certificate
+func getOCSPStatus(cert, issuer *x509.Certificate) (int, error) {
+	req, err := ocsp.CreateRequest(cert, issuer, nil)
+	if err != nil {
+		return ocsp.Unknown, err
+	}
+
+	url := fmt.Sprintf("%s/%s", cert.OCSPServer[0], base64.StdEncoding.EncodeToString(req))
+	res, err := http.Get(url)
+	if err != nil {
+		return ocsp.Unknown, err
+	}
+
+	output, err := io.ReadAll(res.Body)
+	if err != nil {
+		return ocsp.Unknown, err
+	}
+
+	ocspRes, err := ocsp.ParseResponseForCert(output, cert, issuer)
+	if err != nil {
+		return ocsp.Unknown, err
+	}
+
+	switch ocspRes.Status {
+	case ocsp.Revoked:
+		return ocsp.Revoked, nil
+	case ocsp.Good:
+		return ocsp.Good, nil
+	default:
+		return ocsp.Unknown, nil
+	}
+}
+
+// Make a TLS connection with InsecureSkipVerify set to true to get the
+// expiration time and root information of insecure leaf certificates.
+func (p TLSProbe) getInsecureCertInfo() (string, string, time.Duration, error) {
+	conf := &tls.Config{InsecureSkipVerify: true}
+	conn, err := tls.Dial("tcp", p.url+":443", conf)
+	if err != nil {
+		return "", "", 0, err
+	}
+	peers := conn.ConnectionState().PeerCertificates
+	expiryDate := peers[0].NotAfter
+
+	for i := 0; i < len(peers)-1; i++ {
+		certIssuer, issuerSubject := peers[i].Issuer, peers[i+1].Subject
+		if (certIssuer.CommonName != issuerSubject.CommonName) || (certIssuer.Organization[0] != issuerSubject.Organization[0]) {
+			err = fmt.Errorf(reasonToString[issuerVerifyFailed])
+		}
+	}
+
+	root := peers[len(peers)-1].Issuer
+	return root.Organization[0], root.CommonName, time.Until(expiryDate), err
+}
+
+// Check if the root of the leaf's certificate is the one we expect.
+func (p TLSProbe) isRootExpected(rootO, rootCN string) bool {
+	if p.rootOrg == rootO && p.rootCN == rootCN {
+		return true
+	}
+	return false
+}
+
+// Export time to (or since) expiration (in seconds) and reason (with
+// corresponding badOutcome label) to Prometheus.
+func (p TLSProbe) exportMetrics(expiryTime time.Duration, reason reason) {
+	p.certExpiry.WithLabelValues(p.url).Set(expiryTime.Seconds())
+	p.certExpiry.WithLabelValues(p.url, badOutcomeError{reason}.Error()).Set(float64(reason))
+}
+
+// Probe performs the configured TLS protocol. Return true if both root AND
+// response are the expected values, otherwise false. Export time to cert expiry
+// and reason as Prometheus metrics.
+func (p TLSProbe) Probe(timeout time.Duration) (bool, time.Duration) {
+	start := time.Now()
+	conn, secure_err := tls.Dial("tcp", p.url+":443", &tls.Config{})
+	if secure_err != nil {
```

Comment:
This conditional block is basically an entire alternate function -- if we enter this block, we can't exit it, since it ends with an unconditional return. Also, I'm not seeing a reason to attempt two TLS connections, one secure and the backup insecure: in theory, we should only be falling back to insecure when the prober is configured to expect an "invalid" response.

So how about instead we break this into helper functions, and condition which one we call based on the expected result? Something like:
```go
func (p TLSProbe) Probe(timeout time.Duration) (bool, time.Duration) {
  swich p.response:
  case "valid":
    return p.probeUnexpired(timeout, true)
  case "revoked":
    return p.probeUnexpired(timeout, false)
  case "expired":
    return p.probeExpired(timeout)
}
```

Then `probeUnexpired` can use a normal TLS dialer and go on to check OCSP, while the `probeExpired` one can use InsecureSkipVerify and not bother with OCSP.

Note that this may also encourage you to define "valid", "revoked", and "expired" as constants so that they can be used both here and in `validateResponse()`.